internal/controllers/machinedeployment/machinedeployment_rollout_sequence_test.go

```diff
@@ -0,0 +1,876 @@
+				machineDeployment: createMD("v2", 6, 3, 1),
+				machineSets: []*clusterv1.MachineSet{
+					createMS("ms1", "v1", 9),
+					createMS("ms2", "v2", 3),
+				},
+				machineSetMachines: map[string][]*clusterv1.Machine{
+					"ms1": {
+						// "m1", "m2", "m3" already deleted
+						createM("m4", "ms1", "v1"),
+						createM("m5", "ms1", "v1"),
+						createM("m6", "ms1", "v1"),
+						createM("m7", "ms1", "v1"),
+						createM("m8", "ms1", "v1"),
+						createM("m9", "ms1", "v1"),
+						createM("m10", "ms1", "v1"),
+						createM("m11", "ms1", "v1"),
+						createM("m12", "ms1", "v1"),
+					},
+					"ms2": {
+						createM("m13", "ms2", "v2"),
+						createM("m14", "ms2", "v2"),
+						createM("m15", "ms2", "v2"),
+					},
+				},
+				machineUID: 15,
+			},
+			desiredMachineNames:      []string{"m13", "m14", "m15", "m16", "m17", "m18"},
+			maxSurgeBreachToleration: maxSurgeToleration(),
+		},
+		{ // scale out by 3, scale in by 1 (maxSurge > maxUnavailable) + change spec in the middle
+			name:           "Regular rollout, 6 Replicas, maxSurge 3, MaxUnavailable 1, change spec",
+			maxSurge:       3,
+			maxUnavailable: 1,
+			currentScope: &rolloutScope{ // Manually providing a scope simulating a MD with 6 replica in the middle of a rollout, with 3 machines already created in the newMS and 3 still on the oldMS, and then MD spec is changed.
+				machineDeployment: createMD("v3", 6, 3, 1),
+				machineSets: []*clusterv1.MachineSet{
+					createMS("ms1", "v1", 3),
+					createMS("ms2", "v2", 3),
+					createMS("ms3", "v3", 0),
+				},
+				machineSetMachines: map[string][]*clusterv1.Machine{
+					"ms1": {
+						// "m1", "m2", "m3" already deleted
+						createM("m4", "ms1", "v1"),
+						createM("m5", "ms1", "v1"),
+						createM("m6", "ms1", "v1"),
+					},
+					"ms2": {
+						createM("m7", "ms2", "v2"),
+						createM("m8", "ms2", "v2"),
+						createM("m9", "ms2", "v2"),
+					},
+				},
+				machineUID: 9,
+			},
+			desiredMachineNames: []string{"m10", "m11", "m12", "m13", "m14", "m15"}, // NOTE: Machines created before the spec change are deleted
+		},
+	}
+
+	testWithPredictableReconcileOrder := true
+	// TODO(in-place): enable tests with random reconcile order as soon as the issues in reconcileOldMachineSets are fixed
+	testWithRandomReconcileOrderFromConstantSeed := false
+	testWithRandomReconcileOrderFromRandomSeed := false
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			name := tt.name
+
+			if testWithPredictableReconcileOrder {
+				tt.maxIterations = 50
+				tt.randomControllerOrder = false
+				if tt.logAndGoldenFileName == "" {
+					tt.logAndGoldenFileName = strings.ToLower(tt.name)
+				}
+				t.Run("default", func(t *testing.T) {
+					runTestCase(ctx, t, tt, fLogger)
+				})
+			}
+
+			if testWithRandomReconcileOrderFromConstantSeed {
+				tt.maxIterations = 70
```

Comment:
How do we make sure that l.260 does not modify the scope and it has impact on this test case? (similar question for testWithRandomReconcileOrderFromRandomSeed)